aten/src/ATen/native/TensorShape.cpp

```diff
@@ -513,6 +513,40 @@ std::vector<Tensor> chunk(const Tensor& self, int64_t chunks, int64_t dim) {
   }
 }
 
+static void check_tensor_split_dim(const Tensor& self, int64_t dim) {
+  TORCH_CHECK(self.dim() > 0, "expected at least a 1-dimensional tensor");
+  TORCH_CHECK(dim < self.dim(), "expected dim to be less than self.dim()");
+}
+
+std::vector<Tensor> tensor_split(const Tensor& self, int64_t sections, int64_t dim) {
+  check_tensor_split_dim(self, dim);
+  TORCH_CHECK(sections > 0, "number of sections must be larger than 0, got ", sections);
+  std::vector<Tensor> splits(sections);
+  int64_t min_split_size = self.size(dim) / sections;
+  int64_t num_splits_one_extra = self.size(dim) % sections;
+  int64_t start_idx = 0;
+  for (int64_t split_idx = 0; split_idx < sections; split_idx++) {
+    int64_t split_size = (split_idx < num_splits_one_extra) ? (min_split_size + 1) : min_split_size;
+    splits[split_idx] = at::slice(self, dim, start_idx, start_idx + split_size);
+    start_idx += split_size;
+  }
+  return splits;
+}
+
+std::vector<Tensor> tensor_split(const Tensor& self, IntArrayRef indices, int64_t dim) {
+  check_tensor_split_dim(self, dim);
+  int64_t num_indices = indices.size();
+  std::vector<Tensor> splits(num_indices + 1);
+  int64_t start_idx = 0;
+  for (int64_t split_idx = 0; split_idx < num_indices; split_idx++) {
+    int64_t end_idx = indices[split_idx];
+    splits[split_idx] = at::slice(self, dim, start_idx, end_idx);
```

Comment:
What happens when slicing past the end of an array? In NumPy, for example:

```
a = np.array((1, 2 ,3))
np.array_split(a, 6)
: [array([1]),
 array([2]),
 array([3]),
 array([], dtype=int64),
 array([], dtype=int64),
 array([], dtype=int64)]

np.array_split(a, (1, 2, 3, 7, 8))
: [array([1]),
 array([2]),
 array([3]),
 array([], dtype=int64),
 array([], dtype=int64),
 array([], dtype=int64)]
```

Reply:
Yeah, `at::slice` works correctly in this case:
```
>>> torch.tensor_split(torch.tensor([1, 2, 3]), 6)
(tensor([1]), tensor([2]), tensor([3]), tensor([], dtype=torch.int64), tensor([], dtype=torch.int64), tensor([], dtype=torch.int64))
>>> torch.tensor_split(torch.tensor([1, 2, 3]), (1, 2, 3, 7, 8))
(tensor([1]), tensor([2]), tensor([3]), tensor([], dtype=torch.int64), tensor([], dtype=torch.int64), tensor([], dtype=torch.int64))
```